compiler/src/dotty/tools/dotc/transform/CheckUnused.scala

```diff
@@ -0,0 +1,268 @@
+
+  override def isRunnable(using Context): Boolean =
+    ctx.settings.WunusedHas.imports ||
+    ctx.settings.WunusedHas.locals  ||
+    ctx.settings.WunusedHas.privates
+
+  override def run(using Context): Unit =
+    val tree = ctx.compilationUnit.tpdTree
+    val data = UnusedData(ctx)
+    val fresh = ctx.fresh.setProperty(_key, data)
+    traverser.traverse(tree)(using fresh)
+    reportUnused(data.getUnused)
+
+  /**
+   * This traverse is the **main** component of this phase
+   *
+   * It traverse the tree the tree and gather the data in the
+   * corresponding context property
+   */
+  private def traverser = new TreeTraverser {
+    import tpd._
+    import UnusedData.ScopeType
+
+    override def traverse(tree: tpd.Tree)(using Context): Unit = tree match
+      case imp@Import(_, sels) => sels.foreach { s =>
+          ctx.property(_key).foreach(_.registerImport(imp))
+        }
+      case ident: Ident =>
+        val id = ident.symbol.id
+        ctx.property(_key).foreach(_.registerUsed(id))
+        traverseChildren(tree)
+      case sel: Select =>
+        val id = sel.symbol.id
+        ctx.property(_key).foreach(_.registerUsed(id))
+        traverseChildren(tree)
+      case tpd.Block(_,_) =>
+        var prev: ScopeType = ScopeType.Other
+        ctx.property(_key).foreach { ud =>
+          ud.pushScope()
+          prev = ud.currScope
+          ud.currScope = ScopeType.Local
+        }
+        traverseChildren(tree)
+        ctx.property(_key).foreach { ud =>
+          ud.popScope()
+          ud.currScope = prev
+        }
+      case tpd.Template(_,_,_,_) =>
+        var prev: ScopeType = ScopeType.Other
+        ctx.property(_key).foreach { ud =>
+          ud.pushScope()
+          prev = ud.currScope
+          ud.currScope = ScopeType.Template
+        }
+        traverseChildren(tree)
+        ctx.property(_key).foreach { ud =>
+          ud.popScope()
+          ud.currScope = prev
+        }
+      case t:tpd.ValDef =>
+        ctx.property(_key).foreach(_.registerDef(t))
+        traverseChildren(tree)
+      case t:tpd.DefDef =>
+        ctx.property(_key).foreach(_.registerDef(t))
+        traverseChildren(tree)
+      case _ => traverseChildren(tree)
+
+  }
+
+  private def reportUnused(res: UnusedData.UnusedResult)(using Context) =
+    val UnusedData.UnusedResult(imports, locals, privates) = res
+    /* IMPORTS */
+    if ctx.settings.WunusedHas.imports then
+      imports.foreach { s =>
+        report.warning(i"unused import", s.srcPos)
+      }
+    /* LOCAL VAL OR DEF */
+    if ctx.settings.WunusedHas.locals then
+      locals.foreach { s =>
+        report.warning(i"unused local definition", s.srcPos)
+      }
+    /* PRIVATES VAL OR DEF */
+    if ctx.settings.WunusedHas.privates then
+      privates.foreach { s =>
+        report.warning(i"unused private member", s.srcPos)
+      }
+
+end CheckUnused
+
+object CheckUnused:
+  val phaseName: String = "check unused"
```

Comment:
This name should be whitespace-free, e.g. `"checkUnused"`

Reply:
Thank you!
This has been fixed in commit [950e7de](https://github.com/lampepfl/dotty/pull/16157/commits/950e7dee936a3e4b7102564a5eb9cd9ec6f4b37d)